File: cpp/stab/carnelldist.cpp
#include "carnelldist.h"

timestamp abs(timestamp t) {
    return (t>=0) ? t : -t;
}

CarnellDist::CarnellDist(timestamp tau)
{
    exp = new PrecompExp(tau);
}

CarnellDist::~CarnellDist()
{
    delete exp;
}

double CarnellDist::dist(Spikes a, Spikes b)
{
    double d = 0;

    int la = a.length();
    int lb = b.length();

    for (int i = 1; i < la; ++i) {
        for (int j = 1; j < la; ++j) {
            d = d + exp->val(abs(a(i) - a(j)));
        }
    }

    for (int i = 1; i < lb; ++i) {
        for (int j = 1; j < lb; ++j) {
            d = d + exp->val(abs(b(i) - b(j)));
        }
    }

    for (int i = 1; i < la; ++i) {
        for (int j = 1; j < lb; ++j) {
            d = d - 2 * exp->val(abs(a(i) - b(j)));
        }
    }

    return d;
}

File: matlab/distances/carnellDist.m
function d = carnellDist(a, b, tau) 
%CARNELLDIST Computes the Carnell distance between two spike trains
%   a and b must be timestamp vectors
%   tau is a time constant smoothing the distance between nearby spikes

    d = 0;
    
    la = length(a);
    lb = length(b);
    
    for i = 1:la
        for j = 1:la
            d = d + exp(-abs(a(i) - a(j))/tau);
        end
    end
    
    for i = 1:lb
        for j = 1:lb
            d = d + exp(-abs(b(i) - b(j))/tau);
        end
    end
    
    for i = 1:la
        for j = 1:lb
            d = d - 2 * exp(-abs(a(i) - b(j))/tau);
        end
    end
    
    d = max(0, d);% / (la^2 + lb^2);
    
end
    